sandbox/plc4c/examples/hello-subsciptions/src/hello_subscriptions.c

```diff
@@ -0,0 +1,264 @@
+        printf("Preparing a read-request ... ");
+
+        result = plc4c_connection_create_subscription_request(
+            connection, &subscription_request);
+        if (result != OK) {
+          printf("FAILED\n");
+          return -1;
+        }
+        printf("SUCCESS\n");
+
+        printf("Adding cyclic item for 'RANDOM/foo:INTEGER' ... ");
+        result = plc4c_subscription_request_add_cyclic_item(
+            subscription_request, "RANDOM/foo:INTEGER", 500);
+        if (result != OK) {
+          printf("FAILED\n");
+          return -1;
+        }
+        printf("SUCCESS\n");
+
+        // Execute the read-request.
+        printf("Executing a read-request ... ");
+        result = plc4c_subscription_request_execute(
+            subscription_request, &subscription_request_execution);
+        if (result != OK) {
+          printf("FAILED\n");
+          return -1;
+        } else {
+          state = SUBSCRIPTION_REQUEST_SENT;
+        }
+        break;
+      }
+        // Wait until the read-request execution is finished.
+      case SUBSCRIPTION_REQUEST_SENT: {
+        if (plc4c_subscription_request_execution_check_finished_successfully(
+                subscription_request_execution)) {
+          printf("SUCCESS\n");
+          state = SUBSCRIPTION_RESPONSE_RECEIVED;
+        } else if (
+            plc4c_subscription_request_execution_check_finished_with_error(
+                subscription_request_execution)) {
+          printf("FAILED\n");
+          return -1;
+        }
+        break;
+      }
+      case SUBSCRIPTION_RESPONSE_RECEIVED: {
+        // Get the response for the given subscription-request.
+        plc4c_subscription_response *subscription_response =
+            plc4c_subscription_request_execution_get_response(
+                subscription_request_execution);
+        if (subscription_response == NULL) {
+          printf("FAILED (No Response)\n");
+          return -1;
+        }
+
+        // Iterate over all returned items.
+        plc4c_list_element *cur_element =
+            plc4c_utils_list_head(subscription_response->response_items);
+        while (cur_element != NULL) {
+          plc4c_response_subscription_item *subscription_item_item =
+              cur_element->value;
+
+          printf("Value %s (%s):", subscription_item_item->item->name,
+                 plc4c_response_code_to_message(
+                     subscription_item_item->response_code));
+
+          if (subscription_item_item->response_code == OK) {
+            // Save the subscription handles...
+            subscription_handle = subscription_item_item->subscription_handle;
+          }
+
+          cur_element = cur_element->next;
+        }
+
+        // Clean up.
+        plc4c_subscription_response_destroy(subscription_response);
+        plc4c_subscription_request_execution_destroy(
+            subscription_request_execution);
+        plc4c_subscription_request_destroy(subscription_request);
+
```

Comment:
I don't am not sure about separating the states here.  Just as I don't understand all the separation of concerns with the item types.
Even if you add multiple types of items, in the end, you would just check for data.  I am not sure why you would check for poll, check for events.  Just check for data.
If they got a cookie set, they can tell which it was....

But I may be missing something, can you explain a bit?

Reply:
Well the protocols usually support these and they are handled completely different. One is some times polling some times a cyclic push ... some protocols support updates on value changes ... in this the PLC actively sends data if something changes. The third kind are events that are not related to values but to events happening in the PLC program. So they are completely different forms of resources or update patterns.

Reply:
Ok, I can see then why you would care about that when you configured the item.  But why when checking for data?  The caller doesn't need to know do they?

So add_xxx_item()
add_yyy_item()
subscribe()
get_data()
get_data()
get_date().....

when you get the data, you can check the type or which item returned